homeassistant/components/xiaomi_aqara.py

```diff
@@ -25,76 +27,57 @@
-    vol.Optional(ATTR_RINGTONE_VOL): vol.All(vol.Coerce(int),
-                                             vol.Clamp(min=0, max=100))
+SERVICE_SCHEMA_REMOVE_DEVICE = vol.Schema({
+    vol.Required(ATTR_DEVICE_ID):
+        vol.All(cv.string, vol.Length(min=14, max=14))
 })
 
-SERVICE_SCHEMA_REMOVE_DEVICE = XIAOMI_AQARA_SERVICE_SCHEMA.extend({
-    vol.Required(ATTR_DEVICE_ID): vol.All(cv.string,
-                                          vol.Length(min=14, max=14))
+
+GATEWAY_CONFIG = vol.Schema({
+    vol.Optional(CONF_MAC): GW_MAC,
+    vol.Optional(CONF_KEY, default=None):
+        vol.All(cv.string, vol.Length(min=16, max=16)),
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=9898): cv.port,
 })
 
-SERVICE_TO_METHOD = {
-    SERVICE_PLAY_RINGTONE: {'method': 'play_ringtone_service',
-                            'schema': SERVICE_SCHEMA_PLAY_RINGTONE},
-    SERVICE_STOP_RINGTONE: {'method': 'stop_ringtone_service'},
-    SERVICE_ADD_DEVICE: {'method': 'add_device_service'},
-    SERVICE_REMOVE_DEVICE: {'method': 'remove_device_service',
-                            'schema': SERVICE_SCHEMA_REMOVE_DEVICE},
-}
-
-
-def _validate_conf(config):
-    """Validate a list of devices definitions."""
-    res_config = []
-    for gw_conf in config:
-        for _conf in gw_conf.keys():
-            if _conf not in [CONF_MAC, CONF_HOST, CONF_PORT, 'key']:
-                raise vol.Invalid('{} is not a valid config parameter'.
-                                  format(_conf))
-
-        res_gw_conf = {'sid': gw_conf.get(CONF_MAC)}
-        if res_gw_conf['sid'] is not None:
-            res_gw_conf['sid'] = res_gw_conf['sid'].replace(":", "").lower()
-            if len(res_gw_conf['sid']) != 12:
-                raise vol.Invalid('Invalid mac address', gw_conf.get(CONF_MAC))
-        key = gw_conf.get('key')
-
-        if key is None:
-            _LOGGER.warning(
-                'Gateway Key is not provided.'
-                ' Controlling gateway device will not be possible.')
-        elif len(key) != 16:
-            raise vol.Invalid('Invalid key {}.'
-                              ' Key must be 16 characters'.format(key))
-        res_gw_conf['key'] = key
-
-        host = gw_conf.get(CONF_HOST)
-        if host is not None:
-            res_gw_conf[CONF_HOST] = host
-            res_gw_conf['port'] = gw_conf.get(CONF_PORT, 9898)
-
-            _LOGGER.warning(
-                'Static address (%s:%s) of the gateway provided. '
-                'Discovery of this host will be skipped.',
-                res_gw_conf[CONF_HOST], res_gw_conf[CONF_PORT])
-
-        res_config.append(res_gw_conf)
-    return res_config
 
+def _fix_conf_defaults(config):
+    """Update some config defaults."""
+    config['sid'] = config.pop(CONF_MAC, None)
+
+    if config.get(CONF_KEY) is None:
+        _LOGGER.warning(
+            'Key is not provided for gateway %s. Controlling the gateway '
```

Comment:
Is this a warning or should we mark this config as invalid?

Reply:
It's a warning! We want to be able to use the gateway without a proper key.

Reply:
Can we connect if we don't have a key?

Reply:
Yes. We are able the receive measurements. We are unable to control an actor without a key.

Reply:
ooh nice, I didn't realise that. Very cool.